### DISCUSSION TITLE
Wait...what? Using ByVal in parameter of a function CALL?

### DISCUSSION THREAD

**Discussion by mburns08109** [twinbasic/twinbasic#965] [Q&A] [answered]
In this code snippet from David Pineault's article on compacting the current database (found here: https://www.devhut.net/access-vba-compact-the-current-database/), I found this:
` 
Set UIA = New CUIAutomation
    Set Access = UIA.ElementFromHandle(**ByVal** Application.hWndAccessApp)
 
    If Not (Access Is Nothing) Then
        Set Condition = UIA.CreatePropertyCondition(UIA_NamePropertyId, ByVal "Compact and Repair Database")
        Set Element = Access.FindFirst(TreeScope_Subtree, Condition)
`

I know about using ByVal/ByRef on the parameters in a function DEFINITION....but to use them during a function CALL like this??!?
What is this supposed to do? override the actual ByRef parameter or something? not sure I get it at all. What would this save or change, a variable definition (one that can be just a throw-away anyhow)?

I had no idea that VBA supported or tolerated something like this!?

Thinking about it some...I suppose that supporting/allowing this could make some sense from a "supporting compiler optimizations" sort of viewpoint, but...in VBA (which is "compiled"/translated to p-code)? really? ...or is this just something that just sorta works and is basically ignored?

**Comment by wqweto** [comment]
> I had no idea that VBA supported or tolerated something like this!?

Consider this API declaration from API Viewer
```
Private Declare Sub CopyMemory Lib "kernel32" Alias "RtlMoveMemory" (Destination As Any, Source As Any, ByVal Length As Long)
```
For the `As Any` parameter (so called `void *` in IDL) you can use ByVal/ByRef modifier at callsite i.e. `Call CopyMemory(ByRef lPointer, ...)` is different than `Call CopyMemory(ByVal lPointer, ...)` indeed.

**Reply by bclothier** [reply]
Just to provide an example of what wqweto is saying (updated to run on 64-bit VBA):
```
Private Declare PtrSafe Sub CopyMemory Lib "kernel32" Alias "RtlMoveMemory" (Destination As Any, Source As Any, ByVal Length As LongPtr)

Private Type POINT
  X As Long
  Y As Long
End Type
...
Dim p1 As POINT
Dim p2 As POINT

'One way to do it.
CopyMemory p2, p1, LenB(p1)

'Or this, too
Dim pp1 As LongPtr
Dim pp2 As LongPtr

pp1 = VarPtr(p1)
pp2 = VarPtr(p2)

CopyMemory ByVal pp2, ByVal pp1, LenB(p1)
```

If you incorrectly use `ByVal` or `ByRef`, you most likely will crash. In the above, both versions does exactly the same thing. However, there are some APIs that are sensitive to whether you are passing a pointer to something vs. a value. For example, `someFunkyApi ""` may result in a crash since it considers the parameter `""` as "allocated", but not crash when you use `someFunkyApi vbNullString` because the `vbNullString` pass in `0` as the pointer value. Generally, I find it easier to just use `ByVal...As LongPtr` rather than `ByRef...` in API declarations just to avoid the need to remember whether to use `ByVal` at the call site or not.

**Reply by wqweto** [reply]
Btw, in your example if the input UDT (POINT) had a fixed-string member (like font name in LOGFONT) passing `p1` vs `ByVal VarPtr(p1)` would involve vs not-involve Unicode<->ANSI conversion on the fixed string so these would actually differ depending on declared UDT members.

**Comment by EduardoVB** [comment]
> What is this supposed to do? override the actual ByRef parameter or something?

To my understanding, in practice yes. 
Even when the API declaration is ByRef, you can still make a copy from the caller before passing the variable.

The opposite is not possible. I mean, to override a ByVal API declaration with a ByRef calling.

**Reply by mburns08109** [reply]
Hmm...wow, all these years and I still learn new things I never saw before in VB*.

I mean the API calls I can understand (after all, they're sort-of like a Sub/Function Definition, semantically), but somehow, the tidbit about being legit to be decorating the call parameters with ByVal is a new one on me.

**Reply by mburns08109** [reply]
Ok, and I'd read (eons ago) that even if the called function expects a ByRef parameter to be passed, placing a literal value as the parameter turned it into a "de-facto" byBal parameter (as the compiler stored the literal and passed the pointer to it to the sub/function/API call - so no specifying of "ByVal" was even needed for those cases).

**Comment by EduardoVB** [comment]
ByVal means it is a copy. 
Whether it is a copy or not becomes important only when you are going to do something with the original variable (after calling the API).
In case of a literal, the reference is already lost immediately (or there is no reference at all), so ByVal/ByRef becomes irrelevant from your POV.

**Reply by dmrvb** [reply]
I thought that ByVal means that its the actual value that is passed, so I suppose it is a copy, whilst ByRef means its the address of the variable that is passed. Is this correct?  or maybe in reality its more complicated and ByVal passes the address of a copy, but I think its the actual value.,

**Reply by EduardoVB** [reply]
The "actual value" must be the address to a copy, so (to my -limited- understanding) it is the same thing.

**Comment by wqweto** [comment, marked as answer]
Let me make it clear: No you cannot change ByVal/ByRef at callsite unless calling an API declare with `As Any` parameter or calling method on an interface with `void *` parameter from a typelib.

[Here](https://www.vbforums.com/showthread.php?896796-Good-Programming-Practice-About-quot-ByVal-MyObject-quot) is a recent discussion on ByVal vs ByRef.

**Comment by EduardoVB** [comment]
Yes, yes, the parameter in the API needs to be declared As Any.
